File: python/src/revenant/core/pdf/incremental.py
# SPDX-License-Identifier: Apache-2.0
"""PDF structure analysis and incremental update assembly.

Functions for reading existing PDF structure (finding objects, xref offsets)
and building incremental updates (xref tables, trailers, ByteRange patching).

Object-level construction (types, allocation, overrides) is in objects.py.
High-level signing API is in builder.py.
"""

from __future__ import annotations

import re
import zlib

from ...errors import PDFError
from .. import require_pikepdf as _require_pikepdf
from .objects import BYTERANGE_PLACEHOLDER, CMS_HEX_SIZE

# ── PDF structure analysis ───────────────────────────────────────────


def find_root_obj_num(pdf_bytes: bytes) -> tuple[int, int]:
    """Find the catalog /Root object number from the trailer.

    Uses the LAST match -- PDFs with incremental updates may redefine
    /Root in later trailers, and the last one is always authoritative.
    """
    matches = list(re.finditer(rb"/Root\s+(\d+)\s+(\d+)\s+R", pdf_bytes))
    if not matches:
        raise PDFError("Cannot find /Root reference in PDF trailer.")
    m = matches[-1]
    return int(m.group(1)), int(m.group(2))


def find_page_obj_num(
    pdf_bytes: bytes, _root_obj_num: int, page_spec: int | str
) -> tuple[int, float, float, bool, list[str]]:
    """Find the object number of the target page.

    Uses pikepdf only for reading -- never saves the PDF.
    Uses BytesIO for in-memory PDF access (no temp files).
    """
    import io

    from .position import get_page_dimensions, resolve_page_index

    pikepdf = _require_pikepdf()
    with pikepdf.open(io.BytesIO(pdf_bytes)) as pdf:
        page_idx = resolve_page_index(pdf, page_spec)
        page_obj = pdf.pages[page_idx].obj
        page_obj_num = page_obj.objgen[0]
        page_w, page_h = get_page_dimensions(pdf, page_idx)
        # Check if page already has /Annots
        has_annots = "/Annots" in page_obj
        existing_annots: list[str] = []
        if has_annots:
            annots = page_obj["/Annots"]
            for i in range(len(annots)):
                ref = annots[i]
                existing_annots.append(f"{ref.objgen[0]} {ref.objgen[1]} R")
        return page_obj_num, page_w, page_h, has_annots, existing_annots


def find_prev_startxref(pdf_bytes: bytes) -> tuple[int, int, list[str], bool]:
    """Find the last startxref offset, max object number, and trailer entries.

    Returns:
        (prev_xref, max_size, trailer_extra, use_xref_stream) where
        trailer_extra is a list of raw trailer lines to carry forward
        (like /Info and /ID), and use_xref_stream indicates whether the
        PDF uses cross-reference streams (PDF 1.5+).
    """
    # Find the LAST startxref in the file -- PDFs with incremental updates
    # have multiple startxref/%%EOF pairs; the last one is authoritative.
    # The regex is lenient: some PDFs have trailing junk after %%EOF.
    matches = list(re.finditer(rb"startxref\s+(\d+)\s+%%EOF", pdf_bytes))
    if not matches:
        raise PDFError("Cannot find startxref in PDF.")
    prev_xref = int(matches[-1].group(1))

    # Determine /Size using pikepdf, which correctly resolves cross-reference
    # streams, incremental updates, and hybrid-reference files.
    # A regex-based approach fails when a PDF has both a cross-reference stream
    # (with large /Size) and a later traditional trailer (with smaller /Size).
    import io

    pikepdf = _require_pikepdf()
    try:
        with pikepdf.open(io.BytesIO(pdf_bytes)) as pdf:
            max_size = int(pdf.trailer["/Size"])
    except (pikepdf.PdfError, KeyError) as e:
        raise PDFError(f"Cannot determine /Size from PDF trailer: {e}") from e

    # Extract extra trailer entries to carry forward (/Info, /ID, etc.)
    # Per PDF spec, incremental update trailers must contain all entries
    # from the previous trailer (except /Prev and /Size which are updated).
    trailer_extra = _extract_trailer_entries(pdf_bytes)

    # Detect if source PDF uses XRef streams -- incremental updates must
    # use the same format (ISO 32000-1 S7.5.8.4)
    use_xref_stream = _detect_xref_streams(pdf_bytes, prev_xref)

    return prev_xref, max_size, trailer_extra, use_xref_stream


def _extract_trailer_entries(pdf_bytes: bytes) -> list[str]:
    """Extract /Info and /ID entries from the trailer.

    Handles both traditional trailers (``trailer << ... >>``) and
    cross-reference stream PDFs where trailer data is stored in the
    xref stream dictionary.  Falls back to pikepdf when no traditional
    trailer is found, which correctly resolves both formats.
    """
    # Try traditional trailers first (cheaper than pikepdf)
    trailer_extra: list[str] = []
    all_trailers = list(re.finditer(rb"trailer\s*<<(.*?)>>", pdf_bytes, re.DOTALL))
    if all_trailers:
        trailer_content = all_trailers[-1].group(1)
        info_m = re.search(rb"/Info\s+\d+\s+\d+\s+R", trailer_content)
        if info_m:
            trailer_extra.append(info_m.group(0).decode("latin-1"))
        id_m = re.search(rb"/ID\s*\[.*?\]", trailer_content, re.DOTALL)
        if id_m:
            trailer_extra.append(id_m.group(0).decode("latin-1"))
        if trailer_extra:
            return trailer_extra

    # No traditional trailer or no entries found -- use pikepdf to read
    # trailer data from cross-reference streams.
    import io

    pikepdf = _require_pikepdf()
    with pikepdf.open(io.BytesIO(pdf_bytes)) as pdf:
        trailer = pdf.trailer
        if "/Info" in trailer:
            info_obj = trailer["/Info"]
            if isinstance(info_obj, pikepdf.Object) and info_obj.is_indirect:
                trailer_extra.append(f"/Info {info_obj.objgen[0]} {info_obj.objgen[1]} R")
        if "/ID" in trailer:
            id_array = trailer["/ID"]
            trailer_extra.append(f"/ID {id_array.unparse(resolved=True).decode('latin-1')}")
    return trailer_extra


# ── Incremental update assembly ─────────────────────────────────────


def assemble_incremental_update(
    pdf_bytes: bytes,
    raw_objects: list[tuple[bytes, int]],
    new_size: int,
    prev_xref: int,
    root_obj_num: int,
    root_gen: int,
    trailer_extra: list[str],
    use_xref_stream: bool = False,
) -> bytes:
    """Assemble the full PDF with incremental update appended."""
    # Ensure original PDF ends with \n after %%EOF
    base = pdf_bytes
    if not base.endswith(b"\n"):
        base = base + b"\n"

    update_start = len(base)

    # Collect all new objects and their offsets
    objects_raw: list[bytes] = []
    xref_entries: dict[int, int] = {}

    running_offset = update_start
    for raw_bytes, obj_num in raw_objects:
        xref_entries[obj_num] = running_offset
        objects_raw.append(raw_bytes)
        running_offset += len(raw_bytes)

    all_objects = b"".join(objects_raw)

    # Build xref section (table or stream depending on source PDF format)
    xref_offset = update_start + len(all_objects)

    if use_xref_stream:
        xref_obj_num = new_size
        xref_data = build_xref_stream(
            xref_entries=xref_entries,
            prev_xref=prev_xref,
            root_obj_num=root_obj_num,
            root_gen=root_gen,
            trailer_extra=trailer_extra,
            xref_offset=xref_offset,
            xref_obj_num=xref_obj_num,
        )
        return base + all_objects + xref_data

    xref_data = build_xref_and_trailer(
        xref_entries=xref_entries,
        new_size=new_size,
        prev_xref=prev_xref,
        root_obj_num=root_obj_num,
        root_gen=root_gen,
        trailer_extra=trailer_extra,
        xref_offset=xref_offset,
    )

    return base + all_objects + xref_data


def patch_byterange(full_pdf: bytes, original_len: int) -> tuple[bytes, int, int]:
    """Patch the ByteRange placeholder and return (pdf, hex_start, hex_len)."""
    contents_zeros = "0" * CMS_HEX_SIZE
    contents_marker = f"/Contents <{contents_zeros}>".encode("latin-1")

    # Search from where the incremental update starts
    update_start = original_len
    contents_pos = full_pdf.find(contents_marker, update_start)
    if contents_pos == -1:
        raise PDFError("Cannot find Contents placeholder in prepared PDF.")

    hex_start = contents_pos + len(b"/Contents <")
    hex_end = hex_start + CMS_HEX_SIZE

    # Compute ByteRange values
    br_before_len = hex_start
    br_after_start = hex_end + 1  # +1 for closing ">"
    br_after_len = len(full_pdf) - br_after_start

    byterange_value = (
        f"/ByteRange [{0:>10d} {br_before_len:>10d} {br_after_start:>10d} {br_after_len:>10d}]"
    ).encode("latin-1")

    # Patch the ByteRange placeholder in the incremental update only.
    # Use positional replacement to avoid matching placeholders elsewhere
    # in the original PDF (e.g., if the PDF was previously signed).
    br_pos = full_pdf.find(BYTERANGE_PLACEHOLDER, update_start)
    if br_pos == -1:
        raise PDFError("Cannot find ByteRange placeholder in incremental update.")
    full_pdf = full_pdf[:br_pos] + byterange_value + full_pdf[br_pos + len(BYTERANGE_PLACEHOLDER) :]

    # Recalculate hex_start (should be unchanged since placeholder is same width)
    contents_pos2 = full_pdf.find(b"/Contents <", update_start)
    hex_start = contents_pos2 + len(b"/Contents <")

    return full_pdf, hex_start, CMS_HEX_SIZE


# ── Xref table builder ──────────────────────────────────────────────


def build_xref_and_trailer(
    xref_entries: dict[int, int],
    new_size: int,
    prev_xref: int,
    root_obj_num: int,
    root_gen: int,
    trailer_extra: list[str],
    xref_offset: int,
) -> bytes:
    """Build an xref table and trailer for an incremental update.

    Args:
        xref_entries: Mapping of object number to byte offset.
        new_size: Total object count (/Size value).
        prev_xref: Previous xref offset (/Prev value).
        root_obj_num: Catalog object number for /Root reference.
        root_gen: Catalog generation number.
        trailer_extra: Extra trailer entries to carry forward (/Info, /ID).
        xref_offset: Byte offset where this xref table starts.

    Returns:
        Raw bytes of the xref table, trailer, and %%EOF.
    """
    xref_lines = ["xref"]

    if not xref_entries:
        raise PDFError("Cannot build xref table: no objects to reference.")

    # Group consecutive object numbers for compact xref sections
    sorted_nums = sorted(xref_entries.keys())
    groups: list[list[int]] = []
    current_group = [sorted_nums[0]]
    for n in sorted_nums[1:]:
        if n == current_group[-1] + 1:
            current_group.append(n)
        else:
            groups.append(current_group)
            current_group = [n]
    groups.append(current_group)

    for group in groups:
        xref_lines.append(f"{group[0]} {len(group)}")
        # PDF spec S7.5.4: each xref entry is exactly 20 bytes including EOL.
        # Format: "oooooooooo ggggg n\r\n" = 18 chars + \r + \n = 20 bytes.
        # The \r is part of the format string; \n comes from "\n".join().
        xref_lines.extend(f"{xref_entries[obj_num]:010d} 00000 n\r" for obj_num in group)

    # Trailer -- must carry forward all entries from the previous trailer
    # (per PDF spec S7.5.6: entries from previous trailer that are not
    # overridden must be included)
    xref_lines.append("trailer")
    xref_lines.append("<<")
    xref_lines.append(f"  /Size {new_size}")
    xref_lines.append(f"  /Prev {prev_xref}")
    xref_lines.append(f"  /Root {root_obj_num} {root_gen} R")
    xref_lines.extend(f"  {extra}" for extra in trailer_extra)
    xref_lines.append(">>")
    xref_lines.append("startxref")
    xref_lines.append(str(xref_offset))
    xref_lines.append("%%EOF")
    xref_lines.append("")  # trailing newline

    return "\n".join(xref_lines).encode("latin-1")


def build_xref_stream(
    xref_entries: dict[int, int],
    prev_xref: int,
    root_obj_num: int,
    root_gen: int,
    trailer_extra: list[str],
    xref_offset: int,
    xref_obj_num: int,
) -> bytes:
    """Build a cross-reference stream for an incremental update (PDF 1.5+).

    PDFs that use XRef streams require incremental updates to also use
    XRef streams (ISO 32000-1 S7.5.8.4). This replaces both the xref
    table and trailer with a single stream object.

    Args:
        xref_entries: Mapping of object number to byte offset.
        prev_xref: Previous xref offset (/Prev value).
        root_obj_num: Catalog object number for /Root reference.
        root_gen: Catalog generation number.
        trailer_extra: Extra trailer entries to carry forward (/Info, /ID).
        xref_offset: Byte offset where this XRef stream object starts.
        xref_obj_num: Object number for the XRef stream itself.

    Returns:
        Raw bytes of the XRef stream object, startxref, and %%EOF.
    """
    # Include the XRef stream object in its own cross-reference data
    all_entries = dict(xref_entries)
    all_entries[xref_obj_num] = xref_offset

    # /Size = highest object number + 1
    actual_size = xref_obj_num + 1

    # Determine W (column widths for binary entries per ISO 32000-1 Table 17)
    # Column 1: type (1 byte, value 1 = in-use uncompressed)
    # Column 2: byte offset (variable width, big-endian)
    # Column 3: generation number (1 byte, always 0 for new objects)
    max_offset = max(all_entries.values())
    w2 = _bytes_needed(max_offset)

    # Build /Index subsections and binary stream data
    sorted_nums = sorted(all_entries.keys())
    groups = _group_consecutive(sorted_nums)
    index_parts: list[str] = []
    binary_parts = bytearray()

    for group in groups:
        index_parts.append(f"{group[0]} {len(group)}")
        for obj_num in group:
            offset = all_entries[obj_num]
            binary_parts.append(1)  # type 1 = in-use, uncompressed
            binary_parts.extend(offset.to_bytes(w2, byteorder="big"))
            binary_parts.append(0)  # generation = 0

    compressed = zlib.compress(bytes(binary_parts))

    # Build the XRef stream object (replaces both xref table and trailer)
    lines: list[str] = []
    lines.append(f"{xref_obj_num} 0 obj")
    lines.append("<<")
    lines.append("  /Type /XRef")
    lines.append(f"  /Size {actual_size}")
    lines.append(f"  /Prev {prev_xref}")
    lines.append(f"  /Root {root_obj_num} {root_gen} R")
    lines.append(f"  /W [1 {w2} 1]")
    lines.append(f"  /Index [{' '.join(index_parts)}]")
    lines.append("  /Filter /FlateDecode")
    lines.append(f"  /Length {len(compressed)}")
    lines.extend(f"  {extra}" for extra in trailer_extra)
    lines.append(">>")
    lines.append("stream")

    header = ("\n".join(lines) + "\n").encode("latin-1")
    footer = f"\nendstream\nendobj\nstartxref\n{xref_offset}\n%%EOF\n".encode("latin-1")

    return header + compressed + footer


def _detect_xref_streams(pdf_bytes: bytes, startxref_offset: int) -> bool:
    """Detect whether the PDF uses cross-reference streams (PDF 1.5+).

    Checks the bytes at the last startxref offset: if they match an object
    definition (N N obj) rather than ``xref``, the PDF uses XRef streams.
    """
    end = min(startxref_offset + 40, len(pdf_bytes))
    chunk = pdf_bytes[startxref_offset:end]
    return bool(re.match(rb"\s*\d+\s+\d+\s+obj\b", chunk))


def _group_consecutive(sorted_nums: list[int]) -> list[list[int]]:
    """Group sorted numbers into consecutive runs for xref subsections."""
    if not sorted_nums:
        return []
    groups: list[list[int]] = []
    current_group = [sorted_nums[0]]
    for n in sorted_nums[1:]:
        if n == current_group[-1] + 1:
            current_group.append(n)
        else:
            groups.append(current_group)
            current_group = [n]
    groups.append(current_group)
    return groups


def _bytes_needed(value: int) -> int:
    """Minimum bytes needed to represent a non-negative integer in big-endian."""
    if value <= 0xFF:
        return 1
    if value <= 0xFFFF:
        return 2
    if value <= 0xFFFFFF:
        return 3
    return 4

File: rust/crates/revenant-core/src/pdf/incremental.rs
//! PDF structure analysis and incremental-update assembly.
//!
//! Reads the bits of existing PDF structure that a byte-level scan handles best
//! -- the authoritative `/Root` reference, the last `startxref`, and whether the
//! file uses cross-reference streams -- then assembles the incremental update:
//! the new objects, a matching xref (table or stream), and the trailer. The
//! `/Size` and `/Info`+`/ID` carry-forward come from [`super::reader::PdfReader`]
//! (which resolves xref streams correctly); object dictionaries are read there
//! too.

use std::collections::BTreeMap;
use std::sync::LazyLock;

use regex::bytes::Regex;

use super::objects::{deflate, BYTERANGE_PLACEHOLDER, CMS_HEX_SIZE};
use super::reader::ObjRef;
use crate::{Result, RevenantError};

/// The last `/Root N G R` wins: incremental updates may redefine it.
static ROOT_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"/Root\s+(\d+)\s+(\d+)\s+R").expect("valid /Root regex"));
/// Lenient `startxref` match (some PDFs have trailing junk after `%%EOF`).
static STARTXREF_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"startxref\s+(\d+)\s+%%EOF").expect("valid startxref regex"));
/// At the last `startxref` offset, an object definition (rather than `xref`)
/// means the PDF uses cross-reference streams.
static XREF_STREAM_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^\s*\d+\s+\d+\s+obj\b").expect("valid xref-stream regex"));

// ── Structure analysis ──────────────────────────────────────────────────

/// Find the catalog `/Root` object number and generation from the trailer.
///
/// # Errors
///
/// Returns [`RevenantError::Pdf`] if no `/Root` reference is present.
pub fn find_root_obj_num(pdf_bytes: &[u8]) -> Result<ObjRef> {
    let caps = ROOT_RE.captures_iter(pdf_bytes).last().ok_or_else(|| {
        RevenantError::Pdf("Cannot find /Root reference in PDF trailer.".to_owned())
    })?;
    let num = parse_uint::<u32>(&caps[1])?;
    let gen = parse_uint::<u16>(&caps[2])?;
    Ok(ObjRef::new(num, gen))
}

/// Find the byte offset from the last `startxref`.
///
/// # Errors
///
/// Returns [`RevenantError::Pdf`] if no `startxref` is present.
pub fn find_startxref_offset(pdf_bytes: &[u8]) -> Result<usize> {
    let caps = STARTXREF_RE
        .captures_iter(pdf_bytes)
        .last()
        .ok_or_else(|| RevenantError::Pdf("Cannot find startxref in PDF.".to_owned()))?;
    parse_uint::<usize>(&caps[1])
}

/// Detect whether the PDF uses cross-reference streams (PDF 1.5+).
///
/// An incremental update must match the source format (ISO 32000-1 S7.5.8.4).
#[must_use]
pub fn detect_xref_stream(pdf_bytes: &[u8], startxref_offset: usize) -> bool {
    let end = startxref_offset.saturating_add(40).min(pdf_bytes.len());
    let chunk = pdf_bytes.get(startxref_offset..end).unwrap_or(&[]);
    XREF_STREAM_RE.is_match(chunk)
}

// ── Incremental update assembly ─────────────────────────────────────────

/// The pieces of trailer identity carried forward into an incremental update.
#[derive(Debug, Clone)]
pub struct UpdatePlan<'a> {
    /// The `(raw_bytes, obj_num)` objects to append, in order.
    pub raw_objects: &'a [(Vec<u8>, u32)],
    /// The new `/Size` (one past the highest object number).
    pub new_size: u32,
    /// The previous `startxref` offset (`/Prev`).
    pub prev_xref: usize,
    /// The catalog `/Root` reference.
    pub root: ObjRef,
    /// Extra trailer entries to carry forward (`/Info`, `/ID`).
    pub trailer_extra: &'a [String],
    /// Whether to emit a cross-reference stream instead of a table.
    pub use_xref_stream: bool,
}

/// Assemble the full PDF with the incremental update appended.
///
/// The original bytes are preserved exactly (a newline is added only if the file
/// does not already end with one), so the signed byte range covers the original
/// document unchanged.
///
/// # Errors
///
/// Returns [`RevenantError::Pdf`] if there are no objects to reference, or if
/// xref-stream compression fails.
pub fn assemble_incremental_update(pdf_bytes: &[u8], plan: &UpdatePlan<'_>) -> Result<Vec<u8>> {
    let mut base = pdf_bytes.to_vec();
    if !base.ends_with(b"\n") {
        base.push(b'\n');
    }
    let update_start = base.len();

    let mut xref_entries: BTreeMap<u32, usize> = BTreeMap::new();
    let mut running = update_start;
    let mut objects = Vec::new();
    for (raw, obj_num) in plan.raw_objects {
        xref_entries.insert(*obj_num, running);
        objects.extend_from_slice(raw);
        running = running.saturating_add(raw.len());
    }

    let xref_offset = update_start + objects.len();
    let ObjRef {
        num: root_obj_num,
        gen: root_gen,
    } = plan.root;

    let tail = if plan.use_xref_stream {
        build_xref_stream(
            &xref_entries,
            plan.prev_xref,
            root_obj_num,
            root_gen,
            plan.trailer_extra,
            xref_offset,
            plan.new_size,
        )?
    } else {
        build_xref_and_trailer(
            &xref_entries,
            plan.new_size,
            plan.prev_xref,
            root_obj_num,
            root_gen,
            plan.trailer_extra,
            xref_offset,
        )?
    };

    let mut out = base;
    out.extend_from_slice(&objects);
    out.extend_from_slice(&tail);
    Ok(out)
}

/// A PDF prepared for hash-then-sign: the bytes with an empty signature field
/// and a patched `/ByteRange`, plus where the reserved `/Contents` hex
/// placeholder sits so the caller can hash the ByteRange and later splice in the
/// CMS.
#[derive(Debug, Clone)]
pub struct PreparedPdf {
    /// The prepared PDF bytes.
    pub bytes: Vec<u8>,
    /// Byte offset of the reserved `/Contents` hex placeholder.
    pub contents_hex_offset: usize,
    /// Length of the `/Contents` hex placeholder, in bytes.
    pub contents_hex_len: usize,
}

/// Patch the `/ByteRange` placeholder in the incremental update.
///
/// The search starts at `original_len` so it only touches the freshly appended
/// signature, never a placeholder in a previously signed document.
///
/// # Errors
///
/// Returns [`RevenantError::Pdf`] if either placeholder is missing, the
/// `/Contents` field is not the expected zero-filled shape, or the computed
/// ByteRange overflows the fixed-width placeholder.
pub fn patch_byterange(mut full_pdf: Vec<u8>, original_len: usize) -> Result<PreparedPdf> {
    let prefix = b"/Contents <";
    let contents_pos = find_from(&full_pdf, prefix, original_len).ok_or_else(|| {
        RevenantError::Pdf("Cannot find Contents placeholder in prepared PDF.".to_owned())
    })?;
    let hex_start = contents_pos + prefix.len();
    let hex_end = hex_start + CMS_HEX_SIZE;

    // The reserved field must be exactly CMS_HEX_SIZE zeros closed by '>'.
    let is_zero_filled = full_pdf
        .get(hex_start..hex_end)
        .is_some_and(|s| s.iter().all(|&b| b == b'0'));
    if !is_zero_filled || full_pdf.get(hex_end) != Some(&b'>') {
        return Err(RevenantError::Pdf(
            "Contents placeholder is not the expected zero-filled field.".to_owned(),
        ));
    }

    let br_before_len = hex_start;
    let br_after_start = hex_end + 1; // +1 for the closing '>'
    let br_after_len = full_pdf.len() - br_after_start;
    let byterange_value = format!(
        "/ByteRange [{:>10} {:>10} {:>10} {:>10}]",
        0, br_before_len, br_after_start, br_after_len
    );
    if byterange_value.len() != BYTERANGE_PLACEHOLDER.len() {
        return Err(RevenantError::Pdf(format!(
            "ByteRange value ({} bytes) does not fit the fixed placeholder ({} bytes); \
             the document is too large to sign.",
            byterange_value.len(),
            BYTERANGE_PLACEHOLDER.len()
        )));
    }

    let br_pos =
        find_from(&full_pdf, BYTERANGE_PLACEHOLDER.as_bytes(), original_len).ok_or_else(|| {
            RevenantError::Pdf(
                "Cannot find ByteRange placeholder in incremental update.".to_owned(),
            )
        })?;
    // In-place patch: same width, so no offsets shift and hex_start stays valid.
    full_pdf[br_pos..br_pos + byterange_value.len()].copy_from_slice(byterange_value.as_bytes());

    Ok(PreparedPdf {
        bytes: full_pdf,
        contents_hex_offset: hex_start,
        contents_hex_len: CMS_HEX_SIZE,
    })
}

// ── Xref builders ───────────────────────────────────────────────────────

/// Build an xref table and trailer for an incremental update.
///
/// # Errors
///
/// Returns [`RevenantError::Pdf`] if there are no entries to reference.
pub fn build_xref_and_trailer(
    xref_entries: &BTreeMap<u32, usize>,
    new_size: u32,
    prev_xref: usize,
    root_obj_num: u32,
    root_gen: u16,
    trailer_extra: &[String],
    xref_offset: usize,
) -> Result<Vec<u8>> {
    if xref_entries.is_empty() {
        return Err(RevenantError::Pdf(
            "Cannot build xref table: no objects to reference.".to_owned(),
        ));
    }

    let nums: Vec<u32> = xref_entries.keys().copied().collect();
    let mut lines: Vec<String> = vec!["xref".to_owned()];
    for group in group_consecutive(&nums) {
        lines.push(format!("{} {}", group[0], group.len()));
        for obj_num in group {
            let offset = xref_entries[&obj_num];
            // 20-byte entry: "oooooooooo ggggg n\r" + the joining "\n".
            lines.push(format!("{offset:010} 00000 n\r"));
        }
    }

    lines.push("trailer".to_owned());
    lines.push("<<".to_owned());
    lines.push(format!("  /Size {new_size}"));
    lines.push(format!("  /Prev {prev_xref}"));
    lines.push(format!("  /Root {root_obj_num} {root_gen} R"));
    for extra in trailer_extra {
        lines.push(format!("  {extra}"));
    }
    lines.push(">>".to_owned());
    lines.push("startxref".to_owned());
    lines.push(xref_offset.to_string());
    lines.push("%%EOF".to_owned());
    lines.push(String::new()); // trailing newline

    Ok(lines.join("\n").into_bytes())
}

/// Build a cross-reference stream for an incremental update (PDF 1.5+).
///
/// # Errors
///
/// Returns [`RevenantError::Pdf`] if compression fails.
pub fn build_xref_stream(
    xref_entries: &BTreeMap<u32, usize>,
    prev_xref: usize,
    root_obj_num: u32,
    root_gen: u16,
    trailer_extra: &[String],
    xref_offset: usize,
    xref_obj_num: u32,
) -> Result<Vec<u8>> {
    // The xref stream references itself.
    let mut all_entries = xref_entries.clone();
    all_entries.insert(xref_obj_num, xref_offset);

    let actual_size = xref_obj_num + 1;
    let max_offset = all_entries.values().copied().max().unwrap_or(0);
    let w2 = bytes_needed(max_offset);

    let nums: Vec<u32> = all_entries.keys().copied().collect();
    let mut index_parts: Vec<String> = Vec::new();
    let mut binary = Vec::new();
    for group in group_consecutive(&nums) {
        index_parts.push(format!("{} {}", group[0], group.len()));
        for obj_num in group {
            let offset = all_entries[&obj_num];
            binary.push(1u8); // type 1 = in-use, uncompressed
            binary.extend_from_slice(&offset_be_bytes(offset, w2));
            binary.push(0u8); // generation
        }
    }

    let compressed = deflate(&binary)?;

    let mut lines: Vec<String> = vec![
        format!("{xref_obj_num} 0 obj"),
        "<<".to_owned(),
        "  /Type /XRef".to_owned(),
        format!("  /Size {actual_size}"),
        format!("  /Prev {prev_xref}"),
        format!("  /Root {root_obj_num} {root_gen} R"),
        format!("  /W [1 {w2} 1]"),
        format!("  /Index [{}]", index_parts.join(" ")),
        "  /Filter /FlateDecode".to_owned(),
        format!("  /Length {}", compressed.len()),
    ];
    for extra in trailer_extra {
        lines.push(format!("  {extra}"));
    }
    lines.push(">>".to_owned());
    lines.push("stream".to_owned());

    let mut out = lines.join("\n").into_bytes();
    out.push(b'\n');
    out.extend_from_slice(&compressed);
    out.extend_from_slice(
        format!("\nendstream\nendobj\nstartxref\n{xref_offset}\n%%EOF\n").as_bytes(),
    );
    Ok(out)
}

// ── Helpers ─────────────────────────────────────────────────────────────

/// Group sorted object numbers into consecutive runs for xref subsections.
fn group_consecutive(sorted_nums: &[u32]) -> Vec<Vec<u32>> {
    let mut groups: Vec<Vec<u32>> = Vec::new();
    for &n in sorted_nums {
        match groups.last_mut() {
            Some(group) if group.last().copied() == n.checked_sub(1) => group.push(n),
            _ => groups.push(vec![n]),
        }
    }
    groups
}

/// Minimum bytes to hold a byte offset big-endian, capped at 4 (xref-stream
/// offsets are assumed to fit 32 bits -- true for any real PDF).
fn bytes_needed(value: usize) -> usize {
    if value <= 0xFF {
        1
    } else if value <= 0xFFFF {
        2
    } else if value <= 0x00FF_FFFF {
        3
    } else {
        4
    }
}

/// The low `width` big-endian bytes of an offset.
fn offset_be_bytes(offset: usize, width: usize) -> Vec<u8> {
    let full = u64::try_from(offset).unwrap_or(u64::MAX).to_be_bytes();
    full[full.len().saturating_sub(width)..].to_vec()
}

/// Parse an unsigned integer from ASCII digit bytes.
fn parse_uint<T: std::str::FromStr>(bytes: &[u8]) -> Result<T> {
    std::str::from_utf8(bytes)
        .ok()
        .and_then(|s| s.parse::<T>().ok())
        .ok_or_else(|| RevenantError::Pdf("Malformed integer in PDF structure.".to_owned()))
}

/// Find `needle` in `haystack` at or after `from`.
fn find_from(haystack: &[u8], needle: &[u8], from: usize) -> Option<usize> {
    if needle.is_empty() || from >= haystack.len() {
        return None;
    }
    haystack[from..]
        .windows(needle.len())
        .position(|w| w == needle)
        .map(|p| p + from)
}

#[cfg(test)]
mod tests {
    use super::*;

    const BLANK_LETTER: &[u8] = include_bytes!("testdata/blank_letter.pdf");
    const XREF_STREAM: &[u8] = include_bytes!("testdata/blank_letter_xref_stream.pdf");

    #[test]
    fn finds_root_and_startxref() {
        let root = find_root_obj_num(BLANK_LETTER).unwrap();
        assert!(root.num > 0);
        let off = find_startxref_offset(BLANK_LETTER).unwrap();
        assert!(off > 0 && off < BLANK_LETTER.len());
    }

    #[test]
    fn detects_format() {
        let off = find_startxref_offset(BLANK_LETTER).unwrap();
        assert!(!detect_xref_stream(BLANK_LETTER, off));
        let off2 = find_startxref_offset(XREF_STREAM).unwrap();
        assert!(detect_xref_stream(XREF_STREAM, off2));
    }

    #[test]
    fn missing_root_errors() {
        assert!(find_root_obj_num(b"not a pdf").is_err());
        assert!(find_startxref_offset(b"not a pdf").is_err());
    }

    #[test]
    fn groups_consecutive_runs() {
        assert_eq!(
            group_consecutive(&[1, 2, 3, 7, 8, 10]),
            vec![vec![1, 2, 3], vec![7, 8], vec![10]]
        );
        assert!(group_consecutive(&[]).is_empty());
    }

    #[test]
    fn bytes_needed_boundaries() {
        assert_eq!(bytes_needed(0), 1);
        assert_eq!(bytes_needed(255), 1);
        assert_eq!(bytes_needed(256), 2);
        assert_eq!(bytes_needed(65_535), 2);
        assert_eq!(bytes_needed(65_536), 3);
        assert_eq!(bytes_needed(16_777_216), 4);
    }

    #[test]
    fn offset_bytes_are_big_endian() {
        assert_eq!(offset_be_bytes(0x1234, 2), vec![0x12, 0x34]);
        assert_eq!(offset_be_bytes(0xAB, 1), vec![0xAB]);
        assert_eq!(offset_be_bytes(0x01_0203, 3), vec![0x01, 0x02, 0x03]);
    }

    #[test]
    fn xref_table_has_20_byte_entries() {
        let mut entries = BTreeMap::new();
        entries.insert(4u32, 100usize);
        entries.insert(5u32, 250usize);
        let extra = vec!["/Info 2 0 R".to_owned()];
        let raw = build_xref_and_trailer(&entries, 6, 42, 1, 0, &extra, 999).unwrap();
        let s = String::from_utf8(raw).unwrap();
        assert!(s.starts_with("xref\n4 2\n"), "{s}");
        assert!(s.contains("0000000100 00000 n\r\n"), "{s}");
        assert!(s.contains("/Size 6"), "{s}");
        assert!(s.contains("/Prev 42"), "{s}");
        assert!(s.contains("/Root 1 0 R"), "{s}");
        assert!(s.contains("/Info 2 0 R"), "{s}");
        assert!(s.trim_end().ends_with("%%EOF"), "{s}");
    }

    #[test]
    fn empty_xref_errors() {
        let entries = BTreeMap::new();
        assert!(build_xref_and_trailer(&entries, 1, 0, 1, 0, &[], 0).is_err());
    }
}
